### DISCUSSION TITLE
`composer install` only a specific package

### DISCUSSION THREAD

**Discussion by sadamczyk** [composer/composer#11870] [Q&A / Support] [answered]
I'm trying to install a single package based on it's version in the `composer.lock` in an automation script, to avoid having to install all of the project dependencies when all I need is just that single package.
I just tried running `composer install foo/example-package`, but it gives me an error:
`Invalid argument foo/example-package. Use "composer require foo/example-package" instead to add packages to your composer.json.`

Very unhelpfully `composer install --help` lists `packages` as an argument, but the description tells me, it shouldn't be used?
Why is it even listed as an argument then?

```
$ composer install --help
Usage:
  install [options] [--] [<packages>]...
  i

Arguments:
  packages    Should not be provided, use composer require instead to add a given package to composer.json.
```

Using `composer require foo/example-package` doesn't work as well, because 
1. either the installed package version might not actually match the one in the `composer.lock`
2. or requires manual updating if I pinned the version with `composer require foo/example-package:1.2.3`.

Am I missing something, or is this actually not possible, and I will have to live with the drawbacks of either installing everything unnecessarily, using potentially mismatched version, or requiring manual updating?

The only other solution that comes to mind is parsing the `composer.json` or `composer.lock` with something like `grep` and using that as the version string to `composer require`.

I'm grateful for any ideas. :)

**Comment by Seldaek** [comment, marked as answer]
Yeah, we only support dependencies as a set of things. Installing a single dependency is typically meaningless so this is not supported. You can use `composer show --lock --format=json` to extract data and maybe run a require to create a new composer.json with a new vendor dir containing just the one dependency..

**Reply by sadamczyk** [reply]
Alright, I understand. I'll try that then, thanks!

**Reply by dorkster100** [reply]
this is useful, when your composer is meant to run for example from container that is installed using composer. Eg Laravel project that contains Laravel Sail container. You would want to separately first install Sail container, and then from inside that container install rest of things using composer inside container.

**Reply by CapitaineToinon** [reply]
I had the same issue, the best solution I found is to have composer installed globally, then install sail globally too with `composer global require laravel/sail` then in my project's folder, run `cp -r ~/.config/composer/vendor .` to copy the sail files locally. Then, `vendor/bin/sail up -d` can be used normally.
